src/phase0/state_transition/epoch_processing.rs

```diff
@@ -519,6 +539,120 @@ pub fn process_participation_record_updates<
+    Note: ``get_total_balance`` returns ``EFFECTIVE_BALANCE_INCREMENT`` Gwei minimum to avoid divisions by zero.
+     */
+    get_total_balance(
+        state,
+        &get_unslashed_attesting_indices(state, attestations, context)?,
+        context,
+    )
+}
+
+pub fn weigh_justification_and_finalization<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &mut BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    total_active_balance: Gwei,
+    previous_epoch_target_balance: Gwei,
+    current_epoch_target_balance: Gwei,
+    context: &Context,
+) -> Result<(), Error> {
+    let previous_epoch = get_current_epoch(state, context);
+    let current_epoch = get_current_epoch(state, context);
+    let old_previous_justified_checkpoint = state.previous_justified_checkpoint.clone();
+    let old_current_justified_checkpoint = state.current_justified_checkpoint.clone();
+
+    // Process justifications
+    state.previous_justified_checkpoint = state.current_justified_checkpoint.clone();
+    for i in (1..(JUSTIFICATION_BITS_LENGTH - 1)).rev() {
+        let prev = state
+            .justification_bits
+            .get(i - 1)
+            .expect("at least one bit in justification_bits");
+        state.justification_bits.set(i, prev);
+    }
+    state.justification_bits.set(0, false);
+    if previous_epoch_target_balance * 3 >= total_active_balance * 2 {
+        state.current_justified_checkpoint = Checkpoint {
+            epoch: previous_epoch,
+            root: *get_block_root(state, previous_epoch, context)
+                .expect("block root exists for checkpoint"),
+        };
+        state.justification_bits.set(1, true);
+    }
+    if current_epoch_target_balance * 3 >= total_active_balance * 2 {
+        state.current_justified_checkpoint = Checkpoint {
+            epoch: current_epoch,
+            root: *get_block_root(state, current_epoch, context)
+                .expect("block root exists for checkpoint"),
+        };
+        state.justification_bits.set(0, true);
+    }
+
+    // Process finalizations
+    let bits = &state.justification_bits;
+    // The 2nd/3rd/4th most recent epochs are justified, the 2nd using the 4th as source
+    if bits[1..4].all() && old_previous_justified_checkpoint.epoch + 3 == current_epoch {
+        state.finalized_checkpoint = old_previous_justified_checkpoint.clone();
```

Comment:
Here is where the other `clone`s are. I tried to at least reduce their usage by moving `old_previous_justified_checkpoint` instead of cloning within the second `if` statement below.

Technically, couldn't these first two `if`s be combined into one, thus, eliminating the initial `clone`? And the same would go for the 3rd & 4th `if`s

Reply:
I think the compiler will complain unless you do it this way as you would potentially move the data in the first of each pair and then you would not have the data to move in the second of each pair

I don't think you can merge each set of pairs as these are very particular conditions as written and they don't hold if you start breaking up the logic